src/CachedCalls.jl

```diff
@@ -1,5 +1,86 @@
 module CachedCalls
 
-# Write your package code here.
+using FilePathsBase
+using FilePathsBase: /
+using JLSO
+
+const CACHEDCALLS_DIR = home() / ".cachedcalls"
+isdir(CACHEDCALLS_DIR) || mkdir(CACHEDCALLS_DIR)
+
+export @cached_call
+
+"""
+    @cached_call f(args; kwargs)
+
+Caches the result of `f(args; kwargs)` to disk and returns the result. The next time
+`f(args; kwargs)` is called with the same values of `args` and `kwargs` the cached result
+is returned and `f` is not called again.
+
+Restrictions on `f` apply: it must not mutate its arguments or access/mutate globals.
+These assumptions will not be checked and if violated could mean that an incorrect
+result is returned.
+
+Functions are differentiated by name only, meaning that changing the definition and
+rerunning `@cached_call` will return the wrong result.
+"""
+macro cached_call(ex)
+    Meta.isexpr(ex, :call) || error("Invalid use of `@cached_call`")
+
+    func, args, kwargs = _deconstruct(ex)
+    kw_names = first.(kwargs)
+    kw_values = last.(kwargs)
+
+    return quote
+        h = hash([
+            $(esc(func)), # function name
+            $(esc.(args)...), # arg values
+            $(kw_names)..., # kwarg names
+            $(esc.(kw_values)...) # kwarg values
+        ])
+        fname = $(CACHEDCALLS_DIR) / "$(h).jlso"
+        if isfile(fname)
+            return JLSO.load(fname)[:res]
+        else
+            res = $(esc(ex))
+            JLSO.save(fname, :res => res)
+            return res
+        end
+    end
+end
+
+"""
+    _deconstruct(ex)
+
+Deconstruct expression `ex` to a tuple of function name, arguments, and keyword arguments.
+"""
+function _deconstruct(ex)
+    func, fargs = Iterators.peel(ex.args)
+    length(ex.args) == 1 && return func, [], []
+
+    args = filter(subex -> !Meta.isexpr(subex, [:kw, :parameters]), collect(fargs))
+    kwargs = _extract_kwargs(collect(fargs))
+    return func, args, kwargs
+end
+
+"""
+    _extract_kwargs(callargs::AbstractArray{Any})
```

Comment:
This function is very fiddly because the kwargs can be specified in three different ways (see tests) and that results in different expressions that all need to be parsed to the same thing. Particularly annoying is that the `f(;kwarg)` syntactic sugar adds the need for the `keep` kwarg.

I would love to this in a different way